### ISSUE TITLE
Lossy utf-8 string support

### ISSUE THREAD

**Issue by ladislavmacoun** [capnproto/capnproto-rust#314]
The reference c++ library allows users to decide how to approach invalid utf8 encoded text.
for example.

```capnp
struct Message {
    stringField @0 :Text;
}
```

```cpp
...
::capnp::FlatArrayMessageReader reader = ::capnp::FlatArrayMessageReader(
    kj::ArrayPtr<capnp::word>(reinterpret_cast<capnp::word*>(buffer), sz)
);
Message::Reader root = reader.getRoot<Message>();
std::string msg = root.hasStringField() ? root.getStringField() : "";
std::string valid_utf8;
utf8::replace_invalid(msg.begin(), msg.end(), std::back_inserter(valid_utf8));
handle_valid_utf8(std::move(valid_utf8));
```

Here we have a Message with one text stringField, if stringField contains invalid utf8 data, the replace_invalid()[1] method replaces the invalid characters with the `0xfffd`, the same approach which is the standard rust String doing for String::from_utf8_lossy(). We are losing some information but still, this is a better approach than throwing out the Message entirely (depending on the use case ofc.).

I am having a hard time replicating something similar in rust with this library. From what I have tried I am only able to extract a valid UTF-8 &str or error. But I cannot find a way to replace the invalid characters with Unicode replacement characters[2]. 

```rust
let msg = capnp::serialize::read_message_from_flat_slice(
    &mut buf,
    capnp::message::ReaderOptions::new(),
)
.unwrap();
let root = msg.get_root::<http_log_capnp::message::Reader>().unwrap();
if root.has_string_field() {
    match root.get_string_field() {
        Ok(string_field) => println!("valid utf8 {:?}", string_field),
        Err(e) => println!("invalid utf8 {:?}", e),
    }
}; 
 ```
 
 Is there any way how achieve something similar to the c++ example above?

[1]: https://github.com/nemtrif/utfcpp/blob/master/source/utf8/checked.h#L132
[2]: https://www.fileformat.info/info/unicode/char/fffd/index.htm

**Comment by dwrensha** [comment]
Yeah, currently `capnp::text::Reader` is just an alias of Rust's `&str`, so there's not a good way to get a handle on one without already doing the utf8 validation.  A terrible hack you do is to define a second message type
```capnp
struct Message2 {
    stringField @0 :Data;
}
```
and read the message at that type.

I think the real fix will be to make `capnp::test::Reader` a proper wrapper that only optionally does the utf8 validation. That can help with performance too, because not all consumers actually need to get read the value as a`&str`.

**Comment by ladislavmacoun** [comment]
Thank you for the reply.
I will try changing the rust consumer scheme to data, looking at the serialization the only difference between the Text and Data fields in capnp is trailing a null byte so it should be fine. Although, the proper fix you are proposing would be nice!

**Comment by osiewicz** [comment]
> I think the real fix will be to make capnp::text::Reader a proper wrapper that only optionally does the utf8 validation. That can help with performance too, because not all consumers actually need to get read the value as a&str.

That would immensely help my use case too, which involves calling `.get_text` several times. UTF validation does show up in my flamegraphs.

**Comment by ladislavmacoun** [comment]
Sounds good to me!

**Comment by dwrensha** [comment]
@antoinecordelle could you describe more about why you need this feature? In particular, is it for performance reasons, or is it just to be able to recover the data in the event of encountering some non-utf8 bytes? If it's the latter, then there's a possibility that we could get away with a much smaller change than #429. 

I'd also be curious to hear about what @osiewicz is doing. Why does `get_text()` need to be called multiple times, what how would making utf-8 validation lazier help that?

In any case, having more context will help justify whatever decision we make.

**Comment by osiewicz** [comment]
Hey, I've moved on from the project since making that comment. @Shaddy should be able to give more perspective.
IIRC, lazy UTF-8 validation would've helped as in some cases only a subset of all fields of a struct might be read (before dropping it) and UTF-8 showed up as a significant contributor to deserialization time. We've had lots of string fields too.

**Comment by antoinecordelle** [comment]
Sure. In this case, this is mostly to be able to recover the data from Text fields that may have some non-utf8 bytes. Performance reasons can be a nice to have but the main reason was non-utf8 text. One note that differs from the initial discussions in this issue though, is that I'd need to be able to recover the data in a non-lossy way, so not replacing invalid bytes by placeholders. 

I mostly implemented it as such in #429 as that seemed like the change that conceptually made the most sense to me. But happy to hear about the simpler way of achieving that if that can limit the breaking change and not require people to change their code.
